Added `Request` dynamic page

nathanclairmonte opened the pull request:
closes #12 

Did the following things in this PR:

## Request page
- Added the dynamic route `Request` page.
  - Takes the request ID as query params and uses `getServerSideProps` to get the request object from DB using that ID.
  - Shows the listing info (images, title, address, bedroom info etc.) for the listing this request is attached to, along with a "REQUEST TO SUBLET" title (colour changes based on the status of the request), and then the price info for the request.
  - For the price info I just took the same calculations that were being done in the CRA app but that can always be changed. Price calculations update dynamically when the price offer is changed.
  - Also has a "Submit Request" button that is greyed out if the request is past or confirmed. Button isn't hooked up to anything yet tho, I figured we would implement that when we finalize the request logic and all that later on.
- Added an API route `/api/requests/[requestId]/index.js`:
  - This folder structure for the API route means that we call it by calling a fetch to `/api/requests/${requestId}`, the index.js just means that's like the base level route that will be called with that URL structure.
  - For e.g. the route `/api/requests/[requestId]/delete.js` from the last PR is called by fetching to `/api/requests/${requestId}/delete`. Maybe that's like trivial and not worth explaining lol but it confused the hell out of me at first so figured it's worth it to like give a deeper explanation here.
  - Anyways this new API route is super simple just fetches the request based on the given ID
- Added an `IncrementalPriceInput.js` component which is just an input with buttons on either side. The plus/minus buttons will add/subtract $25 to the value and the user can also click into the input and edit manually.
- Added a `ListingItemForSublets` component which is basically the same as the regular `ListingItem` component, but for use on the `Sublets` page. Adds these extra functionalities:
  - The component itself now links to the request page for the request that is passed to it (instead of linking to the listing page like how the `ListingItem` component does)
  - For the active and confirmed tabs, the shown listing will only ever have one request (since a user can only have one active/confirmed request for a given listing) (I think lol, lmk if that's a misunderstanding). Clicking on the listing in those tabs will go straight to that one request.
  - For the past tab, the shown listing can have a bunch of like past/rejected requests. So clicking on the listing itself takes the user to the most recent one, and then they can also click on another one from the accordion to go to that specific request.
  - Will display different price text and price text colours for the price display based on if it's in the active, past or confirmed tab
  - Takes a `showActiveBids` prop which can turn off the active bids display if necessary (for e.g. we don't need it in the confirmed tab)

## Miscellaneous
- Fixed the fetching logic in the `Sublets` page; was getting situations where the requests would fetch before the listings and the page would "finish loading" with nothing to display. Doesn't happen anymore now!
- Fixed the issue on the `ListingItem` component where we were conditionally rendering a `Skeleton` loading component inside of a `<p>` tag. React didn't like that lol so I swapped them around; now the `<p>` tag only renders if it's not loading and vice versa
- Moved the `fetchWithTimeout` function to utils, since I'm using it on the dynamic `Request` page now as well, and we'll prob be using it on any other pages that use `getServerSideProps`. Updated the existing `[listingId].js` page to pull that function from utils too.
- Added a `formatPrice` function which just formats an input number as a price (i.e. adds dollar sign, commas at correct places for thousands, 2 decimal places etc.)
- Moved the status arrays (`ACTIVE_STATUSES` etc.) to a `constants.js` file since I'm also using them on the `Request` page now. Figured if we ever want to add/remove options from those arrays it's better that they're in one place so the changes will propagate throughout the app.

cemgunay reviewed (commented):
Looks great!!! Just adding a couple of minor details... Let me know if you agree with them

### Request Page
- Having the 'submit request' button as a sticky div at the bottom of screen
- Clicking the listing details (and images?) navigates to the listing

IncrementalPriceInput
- Validation so that it cant be zero or more than 4 digits? so (0,9999]
- For past and confirmed requests don't allow them to change the price (disable the field)

### Misc
- Idk if my thing was bugging but if you are logged in as a user who has 0 requests the loading skeletons dont go away in sublets page in any of the tabs
  - Just have it show nothing and say '0 active requests' or whatever the top part says
- How would the delete work in the past requests accordian since now when you click it instead of the delete button showing it takes you to the past request
  - Maybe always show the delete icon next to it?
  - Also maybe add a delete button on the request page itself
 
- When in an active request lets show the current highest offer (Can discuss this later not immedieate these are just thoughts)
  - To do this what would be the most optimal way to do this with mongoDB?
    - Query all active requests everytime and take highest amount
    - Have a isHIghestRequest boolean on each request that gets updated everytime a new request is created or updated?
    - Have a highestRequest field on each listing that gets updated everytime a new request is created or updated?

The updates in utils folder are great and will be hella useful!

nathanclairmonte commented:
Great points!! Agreed with & implemented all of them. Just to recap the changes:

- Submit Request button is sticky at bottom of screen now
- Submit Request button disabled for past and confirmed requests
- Clicking on the listing info or the images navigates over to the listing page
- IncrementalPriceInput can't be less than $25 or bigger than $10,000 (capping it at $9,999 was giving some weird situations with the add/subtract $25 so I just rounded it up by one)
- IncrementalPriceInput disabled for past and confirmed requests
- Fixed that issue where it was loading forever if a user had 0 requests
- Changed past requests accordion to always show delete icon (left the hover animation code there commented out in case we wanna use it later on for bigger screen sizes)
- Added delete button to request page
- Added highest offer info to request page (if it's in the active tab). Just did it the same way we did for the ListingItem components, by making another call to the /api/listingactiverequests route. We can always change to one of those other methods later on, but I think this should be okay!
